### PR TITLE
gadget capabilities: add verdict field

### LINKED ISSUE #912: [RFE] trace capabilities gadget should tell results of capability checks
## Current situation

The trace capabilities gadget shows what capability security checks are triggered by applications but it does not tell the result of the check.

## Impact

- When using the gadget for generating pod security policy, we don't know if the current policy is enough or not.
- The gadget cannot be used for audit.

## Ideal future situation

The trace capabilities gadget shows a new column with the result of the capability check.

## Implementation options

Add a kretprobe on cap_capable().

### PR DESCRIPTION
**Pull request by alban**
# gadget capabilities: add verdict field

The gadget capabilities has a new verdict field. It can be either Allow or Deny.

Fixes #912 

## Testing done

This is tested both in local-gadget with Docker and in Kubernetes.

### local-gadget

```
$ docker run -ti --rm busybox
/ # nice --5 echo ok
nice: setpriority(-5): Permission denied

$ docker run --privileged -ti --rm busybox
/ # nice --5 echo ok
ok
```

```
$ sudo ./local-gadget interactive
» create capabilities trace1
State: Started
» stream trace1 --follow
{"namespace":"default","pod":"kind_sutherland","container":"kind_sutherland","type":"normal","mountnsid":4026532866,"pid":322649,"comm":"nice","capName":"
CAP_SYS_NICE","cap":23,"audit":1,"insetid":"0","verdict":"Deny"}
{"namespace":"default","pod":"sleepy_meninsky","container":"sleepy_meninsky","type":"normal","mountnsid":4026532866,"pid":322879,"comm":"nice","capName":"
CAP_SYS_NICE","cap":23,"audit":1,"insetid":"0","verdict":"Allow"}
```

### Kubernetes

Follow the guide:
```
$ ./kubectl-gadget trace capabilities --selector name=set-priority|grep -v CAP_SYS_ADMIN
NODE             NAMESPACE        POD                            CONTAINER        PID     COMM             UID     CAP  NAME             AUDIT  VERDICT 
aks-agentpool-34376147-vmss000004 default          set-priority-768db6dcf7-rp8gd  set-priority     10158   nice             0       23   CAP_SYS_NICE     1      Allow  
aks-agentpool-34376147-vmss000004 default          set-priority-768db6dcf7-rp8gd  set-priority     10365   nice             0       23   CAP_SYS_NICE     1      Allow  
```

### CODE REVIEW

**Comment by alban** [comment]
I added a patch to implement the flag `--filter-non-audit-checks=true|false`

File: cmd/kubectl-gadget/trace/capabilities.go [hunk 1]
```diff
@@ -59,12 +63,23 @@ func newCapabilitiesCmd() *cobra.Command {
...
+				},
 			}
 
 			return capabilitiesGadget.Run()
 		},
 	}
 
+	cmd.PersistentFlags().BoolVarP(
+		&filterNonAuditChecks,
+		"filter-non-audit-checks",
```
Comment (mauriciovasquezbernal) [inline cmd/kubectl-gadget/trace/capabilities.go:77]: What about shorter names like `only-audit`, `only-audit-checks`? (I personally find it difficult to understand the "filter + non" combination.

**Review by mauriciovasquezbernal — COMMENTED** [review]
Looks good generally speaking. Only concern is about the name of the flag.

File: pkg/gadgets/trace/capabilities/tracer/bpf/capable.bpf.c [hunk 2]
```diff
@@ -67,17 +79,43 @@ int BPF_KPROBE(ig_trace_cap, const struct cred *cred, struct user_namespace *tar
...
+	__u64 pid_tgid;
+	struct args_t *ap;
+	u64 mntns_id;
+	struct task_struct *task;
+	int ret;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	ap = bpf_map_lookup_elem(&start, &pid_tgid);
+	if (!ap)
+		return 0;	/* missed entry */
```
[the remark this answers is not in the document]
Comment (alban) [reply pkg/gadgets/trace/capabilities/tracer/bpf/capable.bpf.c:136]: It is done in the line below but mistakenly in the wrong commit. I'll fix that.

File: cmd/kubectl-gadget/trace/capabilities.go (hunk 1 above)
Comment (alban) [reply cmd/kubectl-gadget/trace/capabilities.go]: Updated to `--only-audit`

**Review by mauriciovasquezbernal — APPROVED** [review]
LGTM!

**Pull request merged**